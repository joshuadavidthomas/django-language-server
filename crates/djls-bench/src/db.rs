use std::io;
use std::sync::Arc;

use camino::Utf8Path;
use camino::Utf8PathBuf;
use djls_semantic::django_builtin_specs;
use djls_semantic::Db as SemanticDb;
use djls_semantic::TagIndex;
use djls_semantic::TagSpecs;
use djls_source::Db as SourceDb;
use djls_source::File;
use djls_source::FxDashMap;
use djls_templates::Db as TemplateDb;
use salsa::Setter;

#[salsa::db]
#[derive(Clone)]
pub struct Db {
    sources: Arc<FxDashMap<Utf8PathBuf, String>>,
    storage: salsa::Storage<Self>,
}

impl Db {
    #[must_use]
    pub fn new() -> Self {
        Self {
            sources: Arc::new(FxDashMap::default()),
            storage: salsa::Storage::default(),
        }
    }

    pub fn file_with_contents(&mut self, path: Utf8PathBuf, contents: &str) -> File {
        self.sources.insert(path.clone(), contents.to_string());
        File::new(self, path, 0)
    }

    pub fn set_file_contents(&mut self, file: File, contents: &str, revision: u64) {
        let path = file.path(self);
        self.sources.insert(path.clone(), contents.to_string());
        file.set_revision(self).to(revision);
    }
}

impl Default for Db {
    fn default() -> Self {
        Self::new()
    }
}

#[salsa::db]
impl salsa::Database for Db {}

#[salsa::db]
impl SourceDb for Db {
    fn create_file(&self, path: &Utf8Path) -> File {
        
        File::new(self, path.to_owned(), 0)
    }

    fn get_file(&self, _path: &Utf8Path) -> Option<File> {
        None
    }

    fn read_file(&self, path: &Utf8Path) -> io::Result<String> {
        Ok(self
            .sources
            .get(path)
            .map(|entry| entry.value().clone())
            .unwrap_or_default())
    }
}

#[salsa::db]
impl TemplateDb for Db {}

#[salsa::db]
impl SemanticDb for Db {
    fn tag_specs(&self) -> TagSpecs {
        django_builtin_specs()
    }

    fn tag_index(&self) -> TagIndex<'_> {
        TagIndex::from_specs(self)
    }

    fn template_dirs(&self) -> Option<Vec<Utf8PathBuf>> {
        None
    }
}
