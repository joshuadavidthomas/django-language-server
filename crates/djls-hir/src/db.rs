use std::sync::Arc;

use djls_templates::Db as TemplateDb;
use djls_workspace::Db as WorkspaceDb;

use crate::specs::TagSpecs;

/// Semantic database trait extending the template and workspace databases
#[salsa::db]
pub trait SemanticDb: TemplateDb + WorkspaceDb {
    /// Get the Django tag specifications for semantic analysis
    fn tag_specs(&self) -> Arc<TagSpecs>;
}