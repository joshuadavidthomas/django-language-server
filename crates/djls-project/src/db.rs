//! Project-specific database trait and queries.
//!
//! This module extends the workspace database trait with project-specific
//! functionality including metadata access and Python environment discovery.
//!
//! ## Architecture
//!
//! Following the Salsa pattern established in workspace and templates crates:
//! - `DjangoProject` is a Salsa input representing external project state
//! - Tracked functions compute derived values (Python env, Django config)
//! - Database trait provides stable configuration (metadata, template tags)

use std::path::Path;
use std::sync::Arc;

use djls_workspace::Db as WorkspaceDb;

use crate::inspector::pool::InspectorPool;
use crate::project::Project;

/// Project-specific database trait extending the workspace database
#[salsa::db]
pub trait Db: WorkspaceDb {
    /// Get the current project (if set)
    fn project(&self) -> Option<Project>;

    /// Get the shared inspector pool for executing Python queries
    fn inspector_pool(&self) -> Arc<InspectorPool>;

    /// Get the project root path if a project is set
    fn project_path(&self) -> Option<&Path> {
        self.project().map(|p| p.root(self).as_path())
    }
}
