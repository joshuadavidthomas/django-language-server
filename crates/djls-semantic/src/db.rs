use std::sync::Arc;

use djls_templates::Db as TemplateDb;
use djls_workspace::Db as WorkspaceDb;

use crate::errors::ValidationError;
use crate::templatetags::TagSpecs;

#[salsa::db]
pub trait Db: TemplateDb + WorkspaceDb {
    /// Get the Django tag specifications for semantic analysis
    fn tag_specs(&self) -> Arc<TagSpecs>;
}

/// Accumulator for validation errors
#[salsa::accumulator]
pub struct ValidationErrorAccumulator(pub ValidationError);
