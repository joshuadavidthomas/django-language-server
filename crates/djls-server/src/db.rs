//! Concrete Salsa database implementation for the Django Language Server.
//!
//! This module provides the concrete [`DjangoDatabase`] that implements all
//! the database traits from workspace, template, and project crates. This follows
//! Ruff's architecture pattern where the concrete database lives at the top level.

use std::path::Path;
use std::path::PathBuf;
use std::sync::Arc;
use std::sync::Mutex;

use dashmap::DashMap;
use djls_project::Db as ProjectDb;
use djls_project::InspectorPool;
use djls_project::Interpreter;
use djls_project::Project;
use djls_semantic::db::Db as SemanticDb;
use djls_semantic::TagSpecs;
use djls_templates::db::Db as TemplateDb;
use djls_workspace::db::Db as WorkspaceDb;
use djls_workspace::db::SourceFile;
use djls_workspace::FileKind;
use djls_workspace::FileSystem;
use salsa::Setter;

/// Concrete Salsa database for the Django Language Server.
///
/// This database implements all the traits from various crates:
/// - [`WorkspaceDb`] for file system access and core operations
/// - [`TemplateDb`] for template parsing and diagnostics
/// - [`ProjectDb`] for project metadata and Python environment
#[salsa::db]
#[derive(Clone)]
pub struct DjangoDatabase {
    /// File system for reading file content (checks buffers first, then disk).
    fs: Arc<dyn FileSystem>,

    /// Maps paths to [`SourceFile`] entities for O(1) lookup.
    files: Arc<DashMap<PathBuf, SourceFile>>,

    /// The single project for this database instance
    project: Arc<Mutex<Option<Project>>>,

    /// Shared inspector pool for executing Python queries
    inspector_pool: Arc<InspectorPool>,

    storage: salsa::Storage<Self>,

    // The logs are only used for testing and demonstrating reuse:
    #[cfg(test)]
    #[allow(dead_code)]
    logs: Arc<Mutex<Option<Vec<String>>>>,
}

#[cfg(test)]
impl Default for DjangoDatabase {
    fn default() -> Self {
        use djls_workspace::InMemoryFileSystem;

        let logs = <Arc<Mutex<Option<Vec<String>>>>>::default();
        Self {
            fs: Arc::new(InMemoryFileSystem::new()),
            files: Arc::new(DashMap::new()),
            project: Arc::new(Mutex::new(None)),
            inspector_pool: Arc::new(InspectorPool::new()),
            storage: salsa::Storage::new(Some(Box::new({
                let logs = logs.clone();
                move |event| {
                    eprintln!("Event: {event:?}");
                    // Log interesting events, if logging is enabled
                    if let Some(logs) = &mut *logs.lock().unwrap() {
                        // only log interesting events
                        if let salsa::EventKind::WillExecute { .. } = event.kind {
                            logs.push(format!("Event: {event:?}"));
                        }
                    }
                }
            }))),
            logs,
        }
    }
}

impl DjangoDatabase {
    /// Set the project for this database instance
    ///
    /// # Panics
    ///
    /// Panics if the project mutex is poisoned.
    pub fn set_project(&self, root: &Path) {
        let interpreter = Interpreter::Auto;
        let django_settings = std::env::var("DJANGO_SETTINGS_MODULE").ok();

        let project = Project::new(self, root.to_path_buf(), interpreter, django_settings);

        *self.project.lock().unwrap() = Some(project);
    }
    /// Create a new [`DjangoDatabase`] with the given file system and file map.
    pub fn new(file_system: Arc<dyn FileSystem>, files: Arc<DashMap<PathBuf, SourceFile>>) -> Self {
        Self {
            fs: file_system,
            files,
            project: Arc::new(Mutex::new(None)),
            inspector_pool: Arc::new(InspectorPool::new()),
            storage: salsa::Storage::new(None),
            #[cfg(test)]
            logs: Arc::new(Mutex::new(None)),
        }
    }

    /// Get an existing [`SourceFile`] for the given path without creating it.
    ///
    /// Returns `Some(SourceFile)` if the file is already tracked, `None` otherwise.
    pub fn get_file(&self, path: &Path) -> Option<SourceFile> {
        self.files.get(path).map(|file_ref| *file_ref)
    }

    /// Get or create a [`SourceFile`] for the given path.
    ///
    /// Files are created with an initial revision of 0 and tracked in the database's
    /// `DashMap`. The `Arc` ensures cheap cloning while maintaining thread safety.
    pub fn get_or_create_file(&mut self, path: &PathBuf) -> SourceFile {
        if let Some(file_ref) = self.files.get(path) {
            return *file_ref;
        }

        // File doesn't exist, so we need to create it
        let kind = FileKind::from_path(path);
        let file = SourceFile::new(self, kind, Arc::from(path.to_string_lossy().as_ref()), 0);

        self.files.insert(path.clone(), file);
        file
    }

    /// Check if a file is being tracked without creating it.
    pub fn has_file(&self, path: &Path) -> bool {
        self.files.contains_key(path)
    }

    /// Touch a file to mark it as modified, triggering re-evaluation of dependent queries.
    ///
    /// Updates the file's revision number to signal that cached query results
    /// depending on this file should be invalidated.
    pub fn touch_file(&mut self, path: &Path) {
        let Some(file_ref) = self.files.get(path) else {
            tracing::debug!("File {} not tracked, skipping touch", path.display());
            return;
        };
        let file = *file_ref;
        drop(file_ref); // Explicitly drop to release the lock

        let current_rev = file.revision(self);
        let new_rev = current_rev + 1;
        file.set_revision(self).to(new_rev);

        tracing::debug!(
            "Touched {}: revision {} -> {}",
            path.display(),
            current_rev,
            new_rev
        );
    }
}

#[salsa::db]
impl salsa::Database for DjangoDatabase {}

#[salsa::db]
impl WorkspaceDb for DjangoDatabase {
    fn fs(&self) -> Arc<dyn FileSystem> {
        self.fs.clone()
    }

    fn read_file_content(&self, path: &Path) -> std::io::Result<String> {
        self.fs.read_to_string(path)
    }
}

#[salsa::db]
impl TemplateDb for DjangoDatabase {}

#[salsa::db]
impl SemanticDb for DjangoDatabase {
    fn tag_specs(&self) -> Arc<TagSpecs> {
        let project_root = if let Some(project) = self.project() {
            project.root(self).clone()
        } else {
            std::env::current_dir().unwrap_or_else(|_| std::path::PathBuf::from("."))
        };

        let tag_specs = if let Ok(settings) = djls_conf::Settings::new(&project_root) {
            TagSpecs::from(&settings)
        } else {
            djls_semantic::django_builtin_specs()
        };

        Arc::new(tag_specs)
    }
}

#[salsa::db]
impl ProjectDb for DjangoDatabase {
    fn project(&self) -> Option<Project> {
        *self.project.lock().unwrap()
    }

    fn inspector_pool(&self) -> Arc<InspectorPool> {
        self.inspector_pool.clone()
    }
}
