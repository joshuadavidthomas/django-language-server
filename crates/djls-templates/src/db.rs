//! Template-specific database trait and queries.
//!
//! This module extends the workspace database trait with template-specific
//! functionality including parsing and diagnostic generation using Salsa accumulators.

use std::sync::Arc;

use djls_workspace::db::SourceFile;
use djls_workspace::Db as WorkspaceDb;
use djls_workspace::FileKind;
use salsa::Accumulator;
use tower_lsp_server::lsp_types;

use crate::ast::AstError;
use crate::ast::LineOffsets;
use crate::ast::Span;
use crate::tagspecs::TagSpecs;
use crate::validation::TagMatcher;
use crate::Ast;
use crate::TemplateError;

/// Diagnostic severity levels
#[derive(Clone, Debug, PartialEq, Eq)]
pub enum DiagnosticSeverity {
    Error,
    Warning,
}

/// Accumulator for template diagnostics - can be a struct directly!
#[salsa::accumulator]
#[derive(Clone, Debug)]
pub struct Diagnostic {
    pub severity: DiagnosticSeverity,
    pub code: String,
    pub message: String,
    pub span: Option<Span>,
}

/// Template-specific database trait extending the workspace database
#[salsa::db]
pub trait Db: WorkspaceDb {
    /// Get the Django tag specifications for template parsing and validation
    fn tag_specs(&self) -> Arc<TagSpecs>;
}

/// Generate a specific error code for an AstError variant
fn ast_error_to_code(error: &AstError) -> String {
    match error {
        AstError::EmptyAst => "DTL-001".to_string(),
        AstError::InvalidTagStructure { .. } => "DTL-002".to_string(),
        AstError::UnbalancedStructure { .. } => "DTL-003".to_string(),
        AstError::InvalidNode { .. } => "DTL-004".to_string(),
        AstError::UnclosedTag { .. } => "DTL-005".to_string(),
        AstError::OrphanedTag { .. } => "DTL-006".to_string(),
        AstError::UnmatchedBlockName { .. } => "DTL-007".to_string(),
        AstError::MissingRequiredArguments { .. } => "DTL-008".to_string(),
        AstError::TooManyArguments { .. } => "DTL-009".to_string(),
    }
}

/// Generate an error code for parser errors
fn parser_error_to_code() -> String {
    "DTL-100".to_string()
}

/// Generate an error code for lexer errors  
fn lexer_error_to_code() -> String {
    "DTL-200".to_string()
}

/// Parse and validate a Django template file.
///
/// This Salsa tracked function parses template files on-demand and caches the results.
/// During parsing and validation, diagnostics are accumulated using the TemplateDiagnostic
/// accumulator. The function returns the parsed AST (or None for non-template files).
///
/// Diagnostics can be retrieved using:
/// ```ignore
/// let diagnostics = 
///     parse_and_validate_template::accumulated::<Diagnostic>(db, file);
/// ```
#[salsa::tracked]
pub fn parse_and_validate_template(db: &dyn Db, file: SourceFile) -> Option<Arc<Ast>> {
    // Only parse template files
    if file.kind(db) != FileKind::Template {
        return None;
    }

    let text_arc = djls_workspace::db::source_text(db, file);
    let text = text_arc.as_ref();

    // Parse the template
    let (ast, parse_errors) = match crate::parse_template(text) {
        Ok((ast, errors)) => (ast, errors),
        Err(err) => {
            // Fatal parse error - accumulate it and return empty AST
            let code = match &err {
                TemplateError::Lexer(_) => lexer_error_to_code(),
                TemplateError::Parser(_) => parser_error_to_code(),
                _ => "DTL-999".to_string(),
            };
            // Extract just the error message without the prefix
            let message = match &err {
                TemplateError::Lexer(msg) => msg.clone(),
                TemplateError::Parser(msg) => msg.clone(),
                _ => err.to_string(),
            };
            Diagnostic {
                severity: DiagnosticSeverity::Error,
                code,
                message,
                span: err.span(),
            }
            .accumulate(db);
            return Some(Arc::new(Ast::default()));
        }
    };

    // Accumulate parse errors
    for error in parse_errors {
        let code = match &error {
            TemplateError::Lexer(_) => lexer_error_to_code(),
            TemplateError::Parser(_) => parser_error_to_code(),
            _ => "DJ999".to_string(),
        };
        // Extract just the error message without the prefix
        let message = match &error {
            TemplateError::Lexer(msg) => msg.clone(),
            TemplateError::Parser(msg) => msg.clone(),
            _ => error.to_string(),
        };
        Diagnostic {
            severity: DiagnosticSeverity::Error,
            code,
            message,
            span: error.span(),
        }
        .accumulate(db);
    }

    // Perform validation and accumulate errors
    let tag_specs = db.tag_specs();
    let (_, validation_errors) = TagMatcher::match_tags(ast.nodelist(), tag_specs);

    for error in validation_errors {
        Diagnostic {
            severity: DiagnosticSeverity::Error,
            code: ast_error_to_code(&error),
            message: error.to_string(),
            span: error.span(),
        }
        .accumulate(db);
    }

    Some(Arc::new(ast))
}

/// Convert a [`Diagnostic`] to an LSP [`Diagnostic`].
///
/// Maps template diagnostics to LSP diagnostics with appropriate
/// severity levels, ranges, and metadata.
pub fn diagnostic_to_lsp(diag: &Diagnostic, line_offsets: &LineOffsets) -> lsp_types::Diagnostic {
    let range = diag
        .span
        .map(|span| {
            // For validation errors (which are Django tags), adjust the span to include delimiters
            let adjusted_span = if diag.code.starts_with("DTL-") && diag.code != "DTL-100" && diag.code != "DTL-200" {
                // Django tags: add delimiter lengths
                // The stored span only includes content length, so add:
                // - 2 for opening {%
                // - 1 for space after {%
                // - content (already in span.length())
                // - 1 for space before %}
                // - 2 for closing %}
                // Total: 6 extra characters
                let start = span.start();
                let length = span.length() + 6;
                Span::new(start, length)
            } else {
                span
            };
            span_to_range(adjusted_span, line_offsets)
        })
        .unwrap_or_default();

    lsp_types::Diagnostic {
        range,
        severity: Some(match diag.severity {
            DiagnosticSeverity::Error => lsp_types::DiagnosticSeverity::ERROR,
            DiagnosticSeverity::Warning => lsp_types::DiagnosticSeverity::WARNING,
        }),
        code: Some(lsp_types::NumberOrString::String(diag.code.clone())),
        code_description: None,
        source: Some("Django Language Server".to_string()),
        message: diag.message.clone(),
        related_information: None,
        tags: None,
        data: None,
    }
}

/// Convert a template [`Span`] to an LSP [`Range`] using line offsets.
#[allow(clippy::cast_possible_truncation)]
fn span_to_range(span: Span, line_offsets: &LineOffsets) -> lsp_types::Range {
    let start_pos = span.start() as usize;
    let end_pos = (span.start() + span.length()) as usize;

    let (start_line, start_char) = line_offsets.position_to_line_col(start_pos);
    let (end_line, end_char) = line_offsets.position_to_line_col(end_pos);

    // Note: These casts are safe in practice as line numbers and character positions
    // in source files won't exceed u32::MAX (4 billion lines/characters)
    lsp_types::Range {
        start: lsp_types::Position {
            line: (start_line - 1) as u32, // LSP is 0-based, LineOffsets is 1-based
            character: start_char as u32,
        },
        end: lsp_types::Position {
            line: (end_line - 1) as u32, // LSP is 0-based, LineOffsets is 1-based
            character: end_char as u32,
        },
    }
}