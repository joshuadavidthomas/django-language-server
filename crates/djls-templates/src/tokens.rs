use djls_source::Span;

use crate::db::Db as TemplateDb;
use crate::nodelist::LineOffsets;

#[derive(Clone, Debug, PartialEq, Hash, salsa::Update)]
pub enum Token<'db> {
    Block {
        content: TokenContent<'db>,
        offset: usize,
    },
    Comment {
        content: TokenContent<'db>,
        offset: usize,
    },
    Error {
        content: TokenContent<'db>,
        offset: usize,
    },
    Eof,
    Newline {
        offset: usize,
    },
    Text {
        content: TokenContent<'db>,
        offset: usize,
    },
    Variable {
        content: TokenContent<'db>,
        offset: usize,
    },
    Whitespace {
        count: usize,
        offset: usize,
    },
}

#[salsa::interned(debug)]
pub struct TokenContent<'db> {
    #[returns(ref)]
    pub text: String,
}

impl<'db> Token<'db> {
    /// Get the content text for content-bearing tokens
    pub fn content(&self, db: &'db dyn TemplateDb) -> String {
        match self {
            Token::Block { content, .. }
            | Token::Comment { content, .. }
            | Token::Error { content, .. }
            | Token::Text { content, .. }
            | Token::Variable { content, .. } => content.text(db).clone(),
            Token::Whitespace { count, .. } => " ".repeat(*count),
            Token::Newline { .. } => "\n".to_string(),
            Token::Eof { .. } => String::new(),
        }
    }

    /// Get the lexeme as it appears in source
    pub fn lexeme(&self, db: &'db dyn TemplateDb) -> String {
        match self {
            Token::Block { content, .. } => format!("{{% {} %}}", content.text(db)),
            Token::Variable { content, .. } => format!("{{{{ {} }}}}", content.text(db)),
            Token::Comment { content, .. } => format!("{{# {} #}}", content.text(db)),
            Token::Text { content, .. } | Token::Error { content, .. } => content.text(db).clone(),
            Token::Whitespace { count, .. } => " ".repeat(*count),
            Token::Newline { .. } => "\n".to_string(),
            Token::Eof { .. } => String::new(),
        }
    }

    pub fn offset(&self) -> Option<u32> {
        match self {
            Token::Block { offset, .. }
            | Token::Comment { offset, .. }
            | Token::Error { offset, .. }
            | Token::Newline { offset, .. }
            | Token::Text { offset, .. }
            | Token::Variable { offset, .. }
            | Token::Whitespace { offset, .. } => {
                Some(u32::try_from(*offset).expect("Offset should fit in u33"))
            }
            Token::Eof { .. } => None,
        }
    }

    /// Get the length of the token content
    pub fn length(&self, db: &'db dyn TemplateDb) -> u32 {
        let len = match self {
            Token::Block { content, .. }
            | Token::Comment { content, .. }
            | Token::Error { content, .. }
            | Token::Text { content, .. }
            | Token::Variable { content, .. } => content.text(db).len(),
            Token::Whitespace { count, .. } => *count,
            Token::Newline { .. } => 1,
            Token::Eof { .. } => 0,
        };
        u32::try_from(len).expect("Token length should fit in u32")
    }
}

#[cfg(test)]
#[derive(Debug, serde::Serialize)]
pub enum TokenSnapshot {
    Block { content: String, offset: usize },
    Comment { content: String, offset: usize },
    Eof,
    Error { content: String, offset: usize },
    Newline { offset: usize },
    Text { content: String, offset: usize },
    Variable { content: String, offset: usize },
    Whitespace { count: usize, offset: usize },
}

#[cfg(test)]
impl<'db> Token<'db> {
    pub fn to_snapshot(&self, db: &'db dyn TemplateDb) -> TokenSnapshot {
        match self {
            Token::Block { content, offset } => TokenSnapshot::Block {
                content: content.text(db).to_string(),
                offset: *offset,
            },
            Token::Comment { content, offset } => TokenSnapshot::Comment {
                content: content.text(db).to_string(),
                offset: *offset,
            },
            Token::Eof => TokenSnapshot::Eof,
            Token::Error { content, offset } => TokenSnapshot::Error {
                content: content.text(db).to_string(),
                offset: *offset,
            },
            Token::Newline { offset } => TokenSnapshot::Newline { offset: *offset },
            Token::Text { content, offset } => TokenSnapshot::Text {
                content: content.text(db).to_string(),
                offset: *offset,
            },
            Token::Variable { content, offset } => TokenSnapshot::Variable {
                content: content.text(db).to_string(),
                offset: *offset,
            },
            Token::Whitespace { count, offset } => TokenSnapshot::Whitespace {
                count: *count,
                offset: *offset,
            },
        }
    }
}

#[cfg(test)]
pub struct TokenSnapshotVec<'db>(pub Vec<Token<'db>>);

#[cfg(test)]
impl TokenSnapshotVec<'_> {
    pub fn to_snapshot(&self, db: &dyn TemplateDb) -> Vec<TokenSnapshot> {
        self.0.iter().map(|t| t.to_snapshot(db)).collect()
    }
}

#[salsa::tracked]
pub struct TokenStream<'db> {
    #[tracked]
    #[returns(ref)]
    pub stream: Vec<Token<'db>>,
    #[tracked]
    #[returns(ref)]
    pub line_offsets: LineOffsets,
}

impl<'db> TokenStream<'db> {
    /// Check if the token stream is empty
    pub fn is_empty(self, db: &'db dyn TemplateDb) -> bool {
        self.stream(db).is_empty()
    }

    /// Get the number of tokens
    pub fn len(self, db: &'db dyn TemplateDb) -> usize {
        self.stream(db).len()
    }
}

pub fn span_from_token(token: &Token<'_>, db: &dyn TemplateDb) -> Span {
    let start = token.offset().unwrap_or(0);
    let length = token.length(db);
    Span::new(start, length)
}
