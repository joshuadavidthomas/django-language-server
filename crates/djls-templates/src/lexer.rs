use djls_source::Span;

use crate::db::Db as TemplateDb;
use crate::tokens::Token;
use crate::tokens::TokenContent;

const BLOCK_TAG_START: &str = "{%";
const BLOCK_TAG_END: &str = "%}";
const VARIABLE_TAG_START: &str = "{{";
const VARIABLE_TAG_END: &str = "}}";
const COMMENT_TAG_START: &str = "{#";
const COMMENT_TAG_END: &str = "#}";

pub struct Lexer<'db> {
    db: &'db dyn TemplateDb,
    source: String,
    start: usize,
    current: usize,
}

impl<'db> Lexer<'db> {
    #[must_use]
    pub fn new(db: &'db dyn TemplateDb, source: &str) -> Self {
        Lexer {
            db,
            source: String::from(source),
            start: 0,
            current: 0,
        }
    }

    pub fn tokenize(&mut self) -> Vec<Token<'db>> {
        let mut tokens = Vec::new();

        while !self.is_at_end() {
            self.start = self.current;

            let token = match self.peek() {
                '{' => match self.peek_next() {
                    '%' => self.lex_django_construct(BLOCK_TAG_END, |content, span| Token::Block {
                        content,
                        span,
                    }),
                    '{' => self.lex_django_construct(VARIABLE_TAG_END, |content, span| {
                        Token::Variable { content, span }
                    }),
                    '#' => self.lex_django_construct(COMMENT_TAG_END, |content, span| {
                        Token::Comment { content, span }
                    }),
                    _ => self.lex_text(),
                },
                c if c.is_whitespace() => self.lex_whitespace(c),
                _ => self.lex_text(),
            };

            tokens.push(token);
        }

        tokens.push(Token::Eof);

        tokens
    }

    fn lex_django_construct(
        &mut self,
        end: &str,
        token_fn: impl FnOnce(TokenContent<'db>, Span) -> Token<'db>,
    ) -> Token<'db> {
        let opening_len = 2;
        let content_start = self.start + opening_len;

        self.consume_n(opening_len);

        match self.consume_until(end) {
            Ok(text) => {
                let content = TokenContent::new(self.db, text);
                let content_end = self.current;
                let span = Span::from_bounds(content_start, content_end);
                self.consume_n(end.len());
                token_fn(content, span)
            }
            Err(err_text) => {
                let content_end = self.current;
                let span = Span::from_bounds(content_start, content_end);
                let content = TokenContent::new(self.db, err_text);
                Token::Error { content, span }
            }
        }
    }

    fn lex_whitespace(&mut self, c: char) -> Token<'db> {
        if c == '\n' || c == '\r' {
            self.consume(); // \r or \n
            if c == '\r' && self.peek() == '\n' {
                self.consume(); // \n of \r\n
            }
            let span = Span::from_bounds(self.start, self.current);
            Token::Newline { span }
        } else {
            self.consume(); // Consume the first whitespace
            while !self.is_at_end() && self.peek().is_whitespace() {
                if self.peek() == '\n' || self.peek() == '\r' {
                    break;
                }
                self.consume();
            }
            let span = Span::from_bounds(self.start, self.current);
            Token::Whitespace { span }
        }
    }

    fn lex_text(&mut self) -> Token<'db> {
        let text_start = self.current;

        while !self.is_at_end() {
            if self.source[self.current..].starts_with(BLOCK_TAG_START)
                || self.source[self.current..].starts_with(VARIABLE_TAG_START)
                || self.source[self.current..].starts_with(COMMENT_TAG_START)
                || self.source[self.current..].starts_with('\n')
            {
                break;
            }
            self.consume();
        }

        let text = &self.source[text_start..self.current];
        let content = TokenContent::new(self.db, text.to_string());
        let span = Span::from_bounds(self.start, self.current);
        Token::Text { content, span }
    }

    #[inline]
    fn peek(&self) -> char {
        self.source[self.current..].chars().next().unwrap_or('\0')
    }

    fn peek_next(&self) -> char {
        let mut chars = self.source[self.current..].chars();
        chars.next(); // Skip current
        chars.next().unwrap_or('\0')
    }

    #[inline]
    fn is_at_end(&self) -> bool {
        self.current >= self.source.len()
    }

    #[inline]
    fn consume(&mut self) {
        if let Some(ch) = self.source[self.current..].chars().next() {
            self.current += ch.len_utf8();
        }
    }

    fn consume_n(&mut self, count: usize) {
        for _ in 0..count {
            self.consume();
        }
    }

    fn consume_until(&mut self, delimiter: &str) -> Result<String, String> {
        let offset = self.current;
        let mut fallback: Option<usize> = None;

        while self.current < self.source.len() {
            if self.source[self.current..].starts_with(delimiter) {
                return Ok(self.source[offset..self.current].to_string());
            }

            if fallback.is_none()
                && (self.source[self.current..].starts_with(BLOCK_TAG_START)
                    || self.source[self.current..].starts_with(VARIABLE_TAG_START)
                    || self.source[self.current..].starts_with(COMMENT_TAG_START))
            {
                fallback = Some(self.current);
            }

            let ch = self.peek();
            if fallback.is_none() && matches!(ch, '\n' | '\r') {
                fallback = Some(self.current);
            }

            self.consume();
        }

        let end = fallback.unwrap_or(self.current);
        let text = self.source[offset..end].to_string();
        self.current = end;
        Err(text)
    }
}
