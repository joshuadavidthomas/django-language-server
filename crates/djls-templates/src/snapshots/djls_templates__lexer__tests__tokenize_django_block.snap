---
source: crates/djls-templates/src/lexer.rs
expression: snapshot
---
- Block:
    content: if user.is_staff
    offset: 3
- Text:
    content: Admin
    offset: 22
- Block:
    content: else
    offset: 30
- Text:
    content: User
    offset: 37
- Block:
    content: endif
    offset: 44
- Eof
